synfig-studio/src/gui/states/state_select.cpp

```diff
@@ -0,0 +1,560 @@
+		vmin[1]=std::min(vmin[1],p[1]);
+		vmax[0]=std::max(vmax[0],p[0]);
+		vmax[1]=std::max(vmax[1],p[1]);
+		positions.push_back(p);
+	}
+	center=(vmin+vmax)*0.5;
+	if((vmin-vmax).mag()<=EPSILON)
+		move_only=true;
+	else
+		move_only=false;
+
+
+	synfig::Vector vect(offset-center);
+	original_angle=Angle::tan(vect[1],vect[0]);
+	original_mag=vect.mag();
+}
+
+bool DuckDrag_Select::end_duck_drag(Duckmatic *duckmatic)
+{
+	if(bad_drag)return false;
+
+	//synfigapp::Action::PassiveGrouper group(get_canvas_interface()->get_instance().get(),_("Rotate Ducks"));
+
+	if(is_moving)
+	{
+		duckmatic->signal_edited_selected_movement_ducks();
+		return true;
+	}
+	else
+	{
+		//probably dont need this idk
+		duckmatic->signal_user_click_selected_ducks(0);
+		return false;
+	}
+}
+
+void DuckDrag_Select::duck_drag(Duckmatic *duckmatic, const synfig::Vector &vector)
+{
+	if (!duckmatic) return;
+
+	if(bad_drag)
+		return;
+
+	// this is quick-hack mostly, so need to check if nothing broken
+	//if (!move_only && !scale && !rotate) return; // nothing to do
+
+	//Override axis lock set in workarea when holding down the shift key
+	if (!move_only && (scale || rotate))
+		duckmatic->set_axis_lock(false);
+
+	synfig::Vector vect;
+	if (move_only || (!scale && !rotate))
+		vect= duckmatic->snap_point_to_grid(vector)-drag_offset+snap;
+	else
+		vect= duckmatic->snap_point_to_grid(vector)-center+snap;
+
+	last_move=vect;
+
+	const DuckList selected_movement_ducks(duckmatic->get_selected_movement_ducks());
+	DuckList::const_iterator iter;
+
+	Time time(duckmatic->get_time());
+	int i;
+	if( move_only || (!scale && !rotate) )
+	{
+		for(i=0,iter=selected_movement_ducks.begin();iter!=selected_movement_ducks.end();++iter,i++)
+		{
+			if((*iter)->get_type()==Duck::TYPE_VERTEX || (*iter)->get_type()==Duck::TYPE_POSITION)
+				(*iter)->set_trans_point(positions[i]+vect, time);
+		}
+		for(i=0,iter=selected_movement_ducks.begin();iter!=selected_movement_ducks.end();++iter,i++)
+		{
+			if((*iter)->get_type()!=Duck::TYPE_VERTEX&&(*iter)->get_type()!=Duck::TYPE_POSITION)
+				(*iter)->set_trans_point(positions[i]+vect, time);
+		}
+	}
+
+	if (rotate)
+	{
+		Angle::deg angle(Angle::tan(vect[1],vect[0]));
+		angle=original_angle-angle;
+		if (constrain)
+		{
+			float degrees = angle.get()/15;
+			angle= Angle::deg (degrees>0?std::floor(degrees)*15:std::ceil(degrees)*15);
+		}
+		Real mag(vect.mag()/original_mag);
+		Real sine(Angle::sin(angle).get());
+		Real cosine(Angle::cos(angle).get());
+
+		for(i=0,iter=selected_movement_ducks.begin();iter!=selected_movement_ducks.end();++iter,i++)
+		{
+			if((*iter)->get_type()!=Duck::TYPE_VERTEX&&(*iter)->get_type()!=Duck::TYPE_POSITION)continue;
+
+			Vector x(positions[i]-center),p;
+
+			p[0]=cosine*x[0]+sine*x[1];
+			p[1]=-sine*x[0]+cosine*x[1];
+			if(scale)p*=mag;
+			p+=center;
+			(*iter)->set_trans_point(p, time);
+		}
+		for(i=0,iter=selected_movement_ducks.begin();iter!=selected_movement_ducks.end();++iter,i++)
```

Comment:
Looks like duplicated code.

Reply:
Oh yes, actually almost everything in `begin/end/duck_drag` is the same except for like 4/5 lines in each. I actually here inherit from the `DuckDrag_Combo` and overload these three methods just to make the same 4/5 changes in each. 

I was actually thinking about just making maybe the base `DuckDrag_Combo` more general to avoid having to overload them again with almost the same implementation. So ok it looks like that would be the best here.

Reply:
done.